File: 0973-k-closest-points-to-origin/0973-k-closest-points-to-origin.cpp
class Solution {
public:
struct cmp{
    bool operator()(const pair<int,int>&a, const pair<int,int>&b){
        if(a.first!=b.first){
            // Min wla fisrst lena hai 
            return b.first>a.first;
        }
        // Min wala hi second lemna hogaaa
        return b.second>a.second;
    }
};

    vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
        int n=points.size();

        // Priority Queue Declaration
        priority_queue<pair<int,int>,vector<pair<int,int>>, cmp> pq;

        for(int i=0;i<n;i++){
            int x=points[i][0];
            int y=points[i][1];

            int dist=x*x+y*y;

            pair<int,int>curr={dist,i};           // Iss trh se raka jaega 

            if(pq.size()<k){
                // Still in the Process ; toh randomly push k elements and move fwd
                pq.push(curr);
                continue;
            }

            if(curr.first>=pq.top().first){
                // skip and move fwd
                continue;
            }
            pq.pop();
            pq.push(curr);
        }

        vector<vector<int>> res;                // Taking a 2D array of the Resultant 

        while(!pq.empty()){
            res.push_back(points[pq.top().second]);          // Res 2D array mein second wley ko push kardo
            pq.pop();                                // ab pop kardo 
        }
        return res;                                // Resultant 2D array return kardo 
    }
};

File: 0973-k-closest-points-to-origin/0973-k-closest-points-to-origin.java
class Solution {
    class Pair{
        int first;
        int second;

        Pair(int first,int second){
            this.first=first;
            this.second=second;
        }
    }

    public int[][] kClosest(int[][] points, int k) {
        int n=points.length;

        // Priority Queue Declaration
        PriorityQueue<Pair> pq=new PriorityQueue<>((a,b)->{
            if(a.first!=b.first){
                // Max wla first lena hai
                return b.first-a.first;
            }
            // Max wala hi second lena hogaaa
            return b.second-a.second;
        });

        for(int i=0;i<n;i++){
            int x=points[i][0];
            int y=points[i][1];

            int dist=x*x+y*y;

            Pair curr=new Pair(dist,i);           // Iss trh se raka jaega

            if(pq.size()<k){
                // Still in the Process ; toh randomly push k elements and move fwd
                pq.add(curr);
                continue;
            }

            if(curr.first>=pq.peek().first){
                // skip and move fwd
                continue;
            }

            pq.poll();
            pq.add(curr);
        }

        int[][] res=new int[k][2];               // Taking a 2D array of the Resultant
        int idx=0;

        while(!pq.isEmpty()){
            res[idx++]=points[pq.peek().second]; // Res 2D array mein second wley ko push kardo
            pq.poll();                            // ab pop kardo
        }

        return res;                               // Resultant 2D array return kardo
    }
}